src/db/models.rs

```diff
@@ -0,0 +1,177 @@
+            .expect("Could not create transaction");
+        insert_into(pushboxv1::table)
+            .values((
+                pushboxv1::user_id.eq(user_id.to_string()),
+                pushboxv1::device_id.eq(device_id.to_string()),
+                pushboxv1::service.eq(service.to_string()),
+                pushboxv1::ttl.eq(ttl),
+                pushboxv1::data.eq(data.clone().into_bytes()),
+            ))
+            .execute(conn)
+            .context(HandlerErrorKind::DBError)?;
+        let record_index = match pushboxv1::table
+            .select(pushboxv1::idx)
+            .order(pushboxv1::idx.desc())
+            .limit(1)
+            .load::<i64>(conn)
+        {
+            Ok(val) => val[0],
+            Err(_) => return Err(HandlerErrorKind::DBError.into()),
+        };
+        t_manager
+            .commit_transaction(conn)
+            .expect("Could not close transaction");
+
+        Ok(record_index)
+    }
+
+    pub fn read_records(
+        conn: &MysqlConnection,
+        user_id: &String,
+        device_id: &String,
+        service: &String,
+        index: i64,
+        limit: i64,
+    ) -> HandlerResult<Vec<Record>> {
+        // flatten into HashMap FromIterator<(K, V)>
+        let mut query = pushboxv1::table
+            .select((
+                pushboxv1::user_id,   // NOTE: load() does not order these, so you should
+                pushboxv1::device_id, // keep them in field order for Record{}
+                pushboxv1::service,
+                pushboxv1::ttl,
+                pushboxv1::idx,
+                pushboxv1::data,
+            ))
+            .into_boxed();
+        query = query
+            .filter(pushboxv1::user_id.eq(user_id.clone()))
+            .filter(pushboxv1::device_id.eq(device_id.clone()))
+            .filter(pushboxv1::service.eq(service.clone()))
+            .filter(pushboxv1::ttl.ge(now_utc()));
+        if index > 0 {
+            query = query.filter(pushboxv1::idx.ge(index));
+        }
+        if limit > 0 {
+            query = query.limit(limit);
+        }
+        Ok(query
+            .order(pushboxv1::idx)
+            .load::<Record>(conn)
+            .context(HandlerErrorKind::DBError)?
+            .into_iter()
+            .collect())
+    }
+
+    pub fn delete(
+        conn: &MysqlConnection,
+        user_id: &String,
+        device_id: &String,
+        service: &String,
+    ) -> HandlerResult<bool> {
+        // boxed deletes are "coming soon"
+        // see https://github.com/diesel-rs/diesel/pull/1534
+        if device_id.len() > 0 {
+            diesel::delete(
+                pushboxv1::table
+                    .filter(pushboxv1::user_id.eq(user_id.clone()))
+                    .filter(pushboxv1::service.eq(service.clone()))
+                    .filter(pushboxv1::device_id.eq(device_id.clone())),
+            ).execute(conn)
+                .context(HandlerErrorKind::DBError)
+                .unwrap();
```

Comment:
It may, but IIUC, the `context()` function captures the error and handles it as a `HandleErrorKind::DBError` <https://boats.gitlab.io/failure/doc/failure/trait.Fail.html#method.context> Although, I admit that it's through the `failure` crate, so there's some voodoo going on there.

Reply:
Right you'll convert the diesel Error into a failure here but you still ultimately end up with a Result that you don't want to unwrap() (panic).

This function already returns a HandlerError so instead it propagate it out via try!: .context(HandlerErrorKind::DBError)?;